Sourcing/TableViewDataSource.swift

```diff
@@ -60,6 +60,9 @@ final public class TableViewDataSource<Object>: NSObject, UITableViewDataSource,
             tableView.reloadRows(at: [indexPath], with: .automatic)
         case .move(let indexPath, let newIndexPath):
             tableView.moveRow(at: indexPath, to: newIndexPath)
+            DispatchQueue.main.async {
```

Comment:
This looks very hacky 

Did an implicit update created by FetchedResultsDataProvider not work

Reply:
You're right. It's a hacky way. Another way to fix it which I have found was add to `FetchedResultsDataProvider` in

```swift
    public func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChange anObject: Any,
                           at indexPath: IndexPath?, for type: NSFetchedResultsChangeType, newIndexPath: IndexPath?) {
        switch type {
        ...
        case .move:
```
lines

```swift
            guard let newObject = self.item(at: newIndexPath) else { fatalError("Object should be not nil") }
            guard let object = self.item(at: indexPath) else { fatalError("Object should be not nil") }
            updates.append(.update(newIndexPath, newObject))
            updates.append(.update(indexPath, object))
```

and then in `TableFetchedResultsDataSource` break batch update in 
```swift
    public func process(updates: [DataProviderUpdate<T>]?) {
        guard let updates = updates else {
            return tableView.reloadData()
        }
        updates.forEach(process)
    }

   private func process(update: DataProviderUpdate<T>) {
        tableView.beginUpdates()
        ...
        tableView.endUpdates()
   } 
```


This way for me is more massy.

Reply:
If you just add to `TableViewDataSource.swift` code:

```swift

    
    private func process(update: DataProviderUpdate<T>) {
        
        switch update {
         ...
        case .move(let indexPath, let newIndexPath):
            tableView.moveRow(at: indexPath, to: newIndexPath)
            tableView.reloadRows(at: [newIndexPath], with: .automatic)
```

you will get error:
```
[error] error: Serious application error.  An exception was caught from the delegate of NSFetchedResultsController during a call to -controllerDidChangeContent:.  attempt to perform an insert and a move to the same index path (<NSIndexPath: 0xc000000000600016> {length = 2, path = 0 - 3}) with userInfo (null)
CoreData: error: Serious application error.  An exception was caught from the delegate of NSFetchedResultsController during a call to -controllerDidChangeContent:.  attempt to perform an insert and a move to the same index path (<NSIndexPath: 0xc000000000600016> {length = 2, path = 0 - 3}) with userInfo (null)
```

> The reloadRowsAtIndexPaths:withRowAnimation: creates an animation for a row being reloaded, and moveRowAtIndexPath:toIndexPath: also creates an animation. Obviously, UITableView can't handle 2 simultaneous animations and throws an exception.

from https://stackoverflow.com/questions/27942659/uitableview-reload-and-move-rows-in-same-batch-update-results-in-attempt-to-cre

Reply:
Would it help if we use tableview.reloadRows with animation .none ?

This could prevent 2 animations.

Reply:
using code like:
```swift
            tableView.reloadRows(at: [indexPath], with: .none)
            tableView.moveRow(at: indexPath, to: newIndexPath)
```

don't fix the bug

Reply:
```
 [error] error: Serious application error.  An exception was caught from the delegate of NSFetchedResultsController during a call to -controllerDidChangeContent:.  attempt to perform a delete and a move from the same index path (<NSIndexPath: 0xc000000000400016> {length = 2, path = 0 - 2}) with userInfo (null)
CoreData: error: Serious application error.  An exception was caught from the delegate of NSFetchedResultsController during a call to -controllerDidChangeContent:.  attempt to perform a delete and a move from the same index path (<NSIndexPath: 0xc000000000400016> {length = 2, path = 0 - 2}) with userInfo (null)
```

Reply:
I thought reloading the newIndexPath would do the trick.

I think if this does not work we should do the following:

1. Go with you Dispatch.async hacky way.
2. Write a good comment which explains this little hack
3. Transfer the hack to `CollectionViewDataSource`
3. Wirte a unit tests for it (I can help with this)

Reply:
Take a look [here](https://github.com/mcudich/HeckelDiff/blob/master/Source/UITableView%2BDiff.swift) how they do it.

Looks like great solution

Reply:
That's what we need

Reply:
Just found another clean solution.

```swift
public func controllerDidChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        dataProviderDidChangeContets(with: updates)
    }
```

Should be:
```swift
public func controllerDidChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        dataProviderDidChangeContets(with: updates)
        let updatesByMoves = updates.map({ (operation: DataProviderUpdate<Object>) -> DataProviderUpdate<Object>? in
            if case .move(_, let newIndexPath) = operation {
                return .update(newIndexPath, object(at: newIndexPath))
            }
            return nil
        }).flatMap { $0 }
        dataProviderDidChangeContets(with: updatesByMoves)
    }
```

Reply:
@lightsprint09 your last solution is the best and works correctly. Use it as a fix.

Reply:
Great. It seems to that this is easy to test as well

Reply:
Do you want to open a PR our should I do it?

Reply:
I'm updating tests right now - give me 20 minutes

Reply:
perfect 👍🏼